### PR TITLE
refactor: separate core.Flow and devkit Flow

### PR DESCRIPTION
**Pull request by elliotzh**
# Description

This PR mainly focus on clean the dependencies from core to devkit. Includes:
1. centralize the usage of `get_used_connection_names`
  1. move cross-flow functionality to `InspectorProxy` in devkit
2. `core.Flow` is not super class of devkit Flow anymore
3. some utils and constants has been moved from devkit to core

This pull request includes a series of changes to the `promptflow` package, focusing on improving the codebase, refactoring the flow execution and inspection logic, and updating the way connections are resolved. The most significant changes involve the creation of a new `ProxyFactory` class, the introduction of `AbstractInspectorProxy` and its subclasses, and the removal of redundant methods.

Refactoring flow execution and inspection logic:

* [`src/promptflow/promptflow/_cli/_pf/_init_entry_generators.py`](diffhunk://#diff-746b4bd84b2bf1ff736d6abe49639a025f3c7c3264f2001e331dafcffe2a5a0dL16-R16): Replaced `FlowOperations` module with `flow_utils` for checking if a flow is executable, and updated the `__init__` method accordingly. [[1]](diffhunk://#diff-746b4bd84b2bf1ff736d6abe49639a025f3c7c3264f2001e331dafcffe2a5a0dL16-R16) [[2]](diffhunk://#diff-746b4bd84b2bf1ff736d6abe49639a025f3c7c3264f2001e331dafcffe2a5a0dL234-R234)
* [`src/promptflow/promptflow/_proxy/_proxy_factory.py`](diffhunk://#diff-936478766b9de290e8791c802585d3932cfed416ae7b50daa76396dacf88cdf5R8-R15): Renamed `ExecutorProxyFactory` to `ProxyFactory` and added a new method `create_inspector_proxy` to create inspector proxies based on the language. [[1]](diffhunk://#diff-936478766b9de290e8791c802585d3932cfed416ae7b50daa76396dacf88cdf5R8-R15) [[2]](diffhunk://#diff-936478766b9de290e8791c802585d3932cfed416ae7b50daa76396dacf88cdf5R54-R64)
* `src/promptflow/promptflow/_proxy/__init__.py`, `src/promptflow/promptflow/_proxy/_base_inspector_proxy.py`, `src/promptflow/promptflow/_proxy/_csharp_inspector_proxy.py`, `src/promptflow/promptflow/_proxy/_python_inspector_proxy.py`: Introduced `AbstractInspectorProxy` and its subclasses `CSharpInspectorProxy` and `PythonInspectorProxy` for inspecting the definition of a Flow. [[1]](diffhunk://#diff-db890b29f2c05ea85ec282550510ac0b0e4761b6be77af142b2e659d4e5a00b0R1-R8) [[2]](diffhunk://#diff-69df25f8ddfbeed622ae5cf13f97f439110586d51262ed1201ce5f3f30f18231R1-R27) [[3]](diffhunk://#diff-3a3b6fcbb15ab97d02280e29888e85dc5c50ca6a78f930442b9fd83a5e219a31R1-R41) [[4]](diffhunk://#diff-a362f804bcd889c4254da0b3249def515e14306ff5dfbf951dff3d8c12b4e7a2R1-R17)

Improvements to connection resolution:

* `src/promptflow/promptflow/_sdk/_submitter/run_submitter.py`, `src/promptflow/promptflow/_sdk/_submitter/test_submitter.py`, `src/promptflow/promptflow/_sdk/_submitter/utils.py`: Updated the connection resolution logic by replacing the `ExecutorProxyFactory` with `ProxyFactory` and removing redundant methods. The `resolve_connections` method now uses the `ProxyFactory` to create an inspector proxy and get used connection names. [[1]](diffhunk://#diff-e3499cf713c66612d91723759de1c5076c22689b2ce2c89678f82de624b57c2cL110-R116) [[2]](diffhunk://#diff-d9dd2c09a149b11eb54626edf065287eeb7b1a28e39719b8dd95377770f64138L248-R225) [[3]](diffhunk://#diff-38c78274ff4a19381a2129bfd86174d225e0d8537b48c80aa8f6feaca25083abL255-R281) [[4]](diffhunk://#diff-38c78274ff4a19381a2129bfd86174d225e0d8537b48c80aa8f6feaca25083abL332-R341)

Codebase simplification:

* [`src/promptflow/promptflow/_core/connection_manager.py`](diffhunk://#diff-9e330c43ad9a3ea8a549e4015cf7972b9778afb4d01f3bdcc2d8f84ea37963acL12-R17): Consolidated import statements for better readability.
* `src/promptflow/promptflow/_sdk/_pf_client.py`, `src/promptflow/promptflow/_sdk/_submitter/run_submitter.py`: Replaced `FlexFlow` import from `entities._eager_flow` to `entities._flow`. [[1]](diffhunk://#diff-c100045cd304dbaa66dd61b7087f78b813ce01e8566d0b823b293bb35755eccfL18-R18) [[2]](diffhunk://#diff-e3499cf713c66612d91723759de1c5076c22689b2ce2c89678f82de624b57c2cL26-R26)
* `src/promptflow/promptflow/_sdk/_submitter/test_submitter.py`, `src/promptflow/promptflow/_sdk/_submitter/utils.py`: Removed unused import statements and methods, and moved the `dump_flow_result` function to `flow_utils`. [[1]](diffhunk://#diff-d9dd2c09a149b11eb54626edf065287eeb7b1a28e39719b8dd95377770f64138R14-R17) [[2]](diffhunk://#diff-d9dd2c09a149b11eb54626edf065287eeb7b1a28e39719b8dd95377770f64138R31-R35) [[3]](diffhunk://#diff-38c78274ff4a19381a2129bfd86174d225e0d8537b48c80aa8f6feaca25083abR28)
* [`src/promptflow/promptflow/_sdk/_utils.py`](diffhunk://#diff-47208ac35b30920275fcd5e55d662647ef360129359bdc77fddd2a2157b6f47eL65): Removed unused `serialize` import from `dataclass_serializer`.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_proxy/__init__.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+
+from ._base_inspector_proxy import AbstractInspectorProxy
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_proxy/__init__.py:5]: add more PR description on this PR

Comment (elliotzh) [reply src/promptflow/promptflow/_proxy/__init__.py:5]: updated

File: src/promptflow/promptflow/_proxy/_base_inspector_proxy.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+from pathlib import Path
+from typing import List
+
+
+class AbstractInspectorProxy:
```
[the remark this answers is not in the document]
Comment (elliotzh) [reply src/promptflow/promptflow/_proxy/_base_inspector_proxy.py:8]: added

**Review by thy09 — DISMISSED** [review]

File: src/promptflow/promptflow/_sdk/entities/_flow/base.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    """A Flow in the context of PromptFlow is a sequence of steps that define a task.
+    Each step in the flow could be a prompt that is sent to a language model, or simply a function task,
+    and the output of one step can be used as the input to the next.
+    Flows can be used to build complex applications with language models.
+
+    Simple Example:
+
+    .. code-block:: python
+
+        from promptflow.core import Flow
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_flow/base.py:140]: update

**Review by wangchao1230 — APPROVED** [review]

**Review by crazygao — APPROVED** [review]

**Pull request merged**